### PR TITLE
crimson/common: allow the tls interrupt_cond to exist when a continuation starts to run

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Release build of seastar invokes continuations synchronously if their predecessor futures
are resolved at the time when those continuations are constructed, so we have to allow
these circumstances to happen
    
Fixes: https://tracker.ceph.com/issues/52275
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
```
218/242 Test #236: unittest-fltree-onode-manager .............Child aborted***Exception:  10.09 sec
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from fltree_onode_manager_test_t
[ RUN      ] fltree_onode_manager_test_t.1_single
Aborting on shard 0.
```

**Comment by tchaikov** [comment]
jenkins test make check

**Review by athanatos — APPROVED** [review]

File: src/crimson/common/interruptible_future.h [hunk 1]
```diff
@@ -117,14 +117,18 @@ auto call_with_interruption_impl(
   // need to be interrupted, return an interruption; otherwise, restore the
   // global "interrupt_cond" with the interruption condition, and go ahead
   // executing the Func.
-  if (!interrupt_cond<InterruptCond> && interrupt_condition) {
+  if (interrupt_condition) {
```
Comment (liu-chunmei) [inline src/crimson/common/interruptible_future.h:120]: do you mean the continuation starts to run before the global "interrupt_cond" is not clear?

**Review by liu-chunmei — APPROVED** [review]

**Comment by cyx1231st** [comment]
Could you also include the revert https://github.com/ceph/ceph/pull/42783 to verify if it works with `unittest-staged-fltree`?

**Comment by xxhdx1985126** [comment]
Hi, everybody, after a second thought, I think there is something we may have to discuss. The issue [52275](https://tracker.ceph.com/issues/52275) is caused by two factors:

1. continuations called synchronously if its predecessor future is resolved when we can `future::then()`;
2. the synchronouly called continuation triggers the event that makes its own continuation chain to interrupt, put it another way, the continuation chain choose to interrupt itself.

I think maybe the second factor needs to be analyzed more carefully. It seems that the major goal of interruptible future is to offer continuation chains to interrupt one another, like the existing pg peering events and seastore transaction confliction events. If a continuation chain wants to interrupt itself, it can actively do so by itself and doesn't need to rely on the interruptible future machanism.

So my question is: do we really need to support this kind of scenarios? I just want to be as careful as possible when modifying a relative basic component since that may affect many parts of the system. Thanks:-)

**Comment by xxhdx1985126** [comment]
> Could you also include the revert #42783 to verify if it works with `unittest-staged-fltree`?

Yeah, consider it done:-)

**Comment by xxhdx1985126** [comment]
> Could you also include the revert #42783 to verify if it works with `unittest-staged-fltree`?

Oh, this pr is based on a relatively old master version, and doesn't have that work around pr.

**Comment by xxhdx1985126** [comment]
> Hi, everybody, after a second thought, I think there is something we may have to discuss. The issue [52275](https://tracker.ceph.com/issues/52275) is caused by two factors:
> 
> 1. continuations called synchronously if its predecessor future is resolved when we can `future::then()`;
> 2. the synchronouly called continuation triggers the event that makes its own continuation chain to interrupt, put it another way, the continuation chain choose to interrupt itself.
> 
> I think maybe the second factor needs to be analyzed more carefully. It seems that the major goal of interruptible future is to offer continuation chains to interrupt one another, like the pg peering events and seastore transaction confliction events. If a continuation chain wants to interrupt itself, it can actively do so by itself and doesn't need to rely on the interruptible future machanism. The reason that we met with this issue is just we need a unit test to make sure the interruptible future works in the onode manager.
> 
> So my question is: do we really need to support this kind of scenarios? I just want to be as careful as possible when modifying a relative basic component since that may affect many parts of the system. Thanks:-)

Per discussion with @tchaikov, it seems that we have to support this kind of scenario, for example, if the interrupt condition is a compose of multiple condtions "conditionA && conditionB", and one continuation chain makes conditionA to be true, and if other transaction makes conditionB to be true in one of its synchronously called continuation, we'll meet with the issue that this pr is trying to deal with.

**Comment by xxhdx1985126** [comment]
> ```
> 218/242 Test #236: unittest-fltree-onode-manager .............Child aborted***Exception:  10.09 sec
> [==========] Running 2 tests from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 2 tests from fltree_onode_manager_test_t
> [ RUN      ] fltree_onode_manager_test_t.1_single
> Aborting on shard 0.
> ```

@tchaikov @athanatos @cyx1231st @rzarzynski @liu-chunmei
Hi, everybody, another discussion:-D

This error is caused by we create a new interrupt condtion within an already interruptible continuation chain, which means there is a different interruptible continuation chain inside another interruptible continuation chain. To support this scenario we probably need to implement some kind of interrupt_cond stack inside the interruptible future.

Again, do we intend to support this scenario? Thanks:-)

**Review by tchaikov — APPROVED** [review]

**Comment by cyx1231st** [comment]
IIUC our conclusion is to support both:
> 1. continuations called synchronously if its predecessor future is resolved when we can future::then();
> 2. the synchronouly called continuation triggers the event that makes its own continuation chain to interrupt, put it another way, the continuation chain choose to interrupt itself.

But won't support nested interruption condition right now.

**Pull request merged**